src/dynimage.rs

```diff
@@ -1177,6 +1187,52 @@ fn image_to_bytes(image: &DynamicImage) -> Vec<u8> {
     }
 }
 
+fn image_into_bytes(image: DynamicImage) -> Vec<u8> {
+    fn cast_vec(mut src: Vec<u16>) -> Vec<u8> {
+        unsafe {
+            let len = 2 * src.len();
+            let cap = 2 * src.capacity();
+            let ptr = src.as_mut_ptr();
+            std::mem::forget(src);
+            Vec::from_raw_parts(ptr as *mut u8, len, cap)
```

Comment:
This is unsound. [In particular](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.from_raw_parts):

> * T needs to have the same size and alignment as what ptr was allocated with. (T having a less strict alignment is not sufficient, the alignment really needs to be equal to satsify the dealloc requirement that memory must be allocated and deallocated with the same layout.)

The size with which it was allocated is `2` and the size you intend to construct is `1`.

Reply:
Ah, that's a good catch. I mistakenly believed that the less strict alignment requirement allowed it to be safe. 

This was the primary use case that I had :(

Perhaps a signature of `into_bytes(self) -> Result<Vec<u8>, Self>` would be suitable? I think that would allow something like `img.into_bytes().map_err(|this| this.to_bytes())` so that the user can get the free conversion when possible and have the copy as a fallback.

Reply:
An alternative would be to use something similar to `take` here.

```rust
fn take_bytes(&mut self) -> Result<Vec<u8>, NotBytesError>
```

That is, on success we leave an empty image and on failure we return some potentially descriptive error. In particular that would allow `From<NotBytesError> for ImageError` as well as a proper `Debug`/`Display` impl which would be more ergonomic in some situations that want to use `?` operator.

Reply:
Wouldn't this have to allocate a new buffer for `self`'s storage?

Reply:
I'm not sure this makes sense to add to DynamicImage if it is going to fail for half the formats and is relatively straightforward to implement in user code.

Reply:
> Wouldn't this have to allocate a new buffer for selfs storage?

No, `Vec::new` is basically free and returns basically a constant.

Reply:
> And is relatively straightforward to implement in user code.

Is it, though? It's an operation that takes the memory from all representations that use bytes as samples. Since the enum is non-exhaustive you can't possibly implement this downstream for all future possibilities. And even for the current ones it's somewhat tedious.

Reply:
> > Wouldn't this have to allocate a new buffer for selfs storage?
> 
> No, `Vec::new` is basically free and returns basically a constant. The trick would be to really empty the buffer such that it has size 0×0.

Yes, `Vec::new` is free, but wouldn't `set_pixel` and `get_pixel` panic if the buffer/vec is empty? In other words, is replacing the internal buffer with a zero-length vec still considered "valid" ?